File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/zebra-puzzle/zebra_puzzle.go
package zebra


type Solution struct {
	DrinksWater string
	OwnsZebra   string
}

type House struct {
	color      string
	national   string
	pet        string
	drink      string
	cigarette  string
}

var (
	nationals = []string{"Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"}
	colors    = []string{"red", "green", "ivory", "yellow", "blue"}
	pets      = []string{"dog", "snails", "fox", "horse", "zebra"}
	drinks    = []string{"coffee", "tea", "milk", "orange juice", "water"}
	cigarettes = []string{"Old Gold", "Kools", "Chesterfields", "Lucky Strike", "Parliaments"}
)

func contains(s []string, str string) bool {
	for _, v := range s {
		if v == str {
			return true
		}
	}
	return false
}

func allUnique(items []string) bool {
	seen := make(map[string]bool)
	for _, item := range items {
		if seen[item] {
			return false
		}
		seen[item] = true
	}
	return true
}

func validateConstraints(houses []House) bool {
	var positions struct {
		nat, col, pet, drink, cig []string
	}
	
	for _, h := range houses {
		positions.nat = append(positions.nat, h.national)
		positions.col = append(positions.col, h.color)
		positions.pet = append(positions.pet, h.pet)
		positions.drink = append(positions.drink, h.drink)
		positions.cig = append(positions.cig, h.cigarette)
	}
	
	if !allUnique(positions.nat) || !allUnique(positions.col) ||
		!allUnique(positions.pet) || !allUnique(positions.drink) || !allUnique(positions.cig) {
		return false
	}

	for i, h := range houses {
		// Known constraints
		if h.national == "Englishman" && h.color != "red" {
			return false
		}
		if h.national == "Spaniard" && h.pet != "dog" {
			return false
		}
		if h.drink == "coffee" && h.color != "green" {
			return false
		}
		if h.national == "Ukrainian" && h.drink != "tea" {
			return false
		}
		if h.cigarette == "Old Gold" && h.pet != "snails" {
			return false
		}
		if h.cigarette == "Kools" && h.color != "yellow" {
			return false
		}
		if h.cigarette == "Lucky Strike" && h.drink != "orange juice" {
			return false
		}
		if h.national == "Japanese" && h.cigarette != "Parliaments" {
			return false
		}
		if i == 2 && h.drink != "milk" {
			return false
		}
		if i == 0 && h.national != "Norwegian" {
			return false
		}
	}

	// Cross-house constraints
	greenIdx := -1
	ivoryIdx := -1
	for i, h := range houses {
		if h.color == "green" {
			greenIdx = i
		}
		if h.color == "ivory" {
			ivoryIdx = i
		}
	}
	if greenIdx != ivoryIdx+1 {
		return false
	}

	// Norwegian next to blue
	if houses[0].color == "blue" || (houses[1].color != "blue" && len(houses) > 1) {
		return false
	}

	// Chesterfields next to fox
	chesterfieldIdx := -1
	foxIdx := -1
	for i, h := range houses {
		if h.cigarette == "Chesterfields" {
			chesterfieldIdx = i
		}
		if h.pet == "fox" {
			foxIdx = i
		}
	}
	if abs(chesterfieldIdx-foxIdx) != 1 {
		return false
	}

	// Kools next to horse
	koolsIdx := -1
	horseIdx := -1
	for i, h := range houses {
		if h.cigarette == "Kools" {
			koolsIdx = i
		}
		if h.pet == "horse" {
			horseIdx = i
		}
	}
	if abs(koolsIdx-horseIdx) != 1 {
		return false
	}

	return true
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func backtrack(houses []House, usedN, usedC, usedP, usedD, usedS map[string]bool) []House {
	if len(houses) == 5 {
		if validateConstraints(houses) {
			return houses
		}
		return nil
	}

	i := len(houses)
	for _, nat := range nationals {
		if i == 0 && nat != "Norwegian" {
			continue
		}
		if usedN[nat] {
			continue
		}
		for _, col := range colors {
			if i == 1 && col != "blue" {
				continue
			}
			if usedC[col] {
				continue
			}
			for _, pet := range pets {
				if usedP[pet] {
					continue
				}
				for _, drink := range drinks {
					if i == 2 && drink != "milk" {
						continue
					}
					if usedD[drink] {
						continue
					}
					for _, cig := range cigarettes {
						if usedS[cig] {
							continue
						}

						newHouse := House{
							national:  nat,
							color:    col,
							pet:      pet,
							drink:    drink,
							cigarette: cig,
						}

						newUsedN := cloneAndAdd(usedN, nat)
						newUsedC := cloneAndAdd(usedC, col)
						newUsedP := cloneAndAdd(usedP, pet)
						newUsedD := cloneAndAdd(usedD, drink)
						newUsedS := cloneAndAdd(usedS, cig)

						result := backtrack(append(houses, newHouse), newUsedN, newUsedC, newUsedP, newUsedD, newUsedS)
						if result != nil {
							return result
						}
					}
				}
			}
		}
	}
	return nil
}

func cloneAndAdd(m map[string]bool, key string) map[string]bool {
	newMap := make(map[string]bool)
	for k, v := range m {
		newMap[k] = v
	}
	newMap[key] = true
	return newMap
}

func SolvePuzzle() Solution {
	houses := backtrack(nil, make(map[string]bool), make(map[string]bool), 
		make(map[string]bool), make(map[string]bool), make(map[string]bool))
		
	var waterDrinker, zebraOwner string
	for _, h := range houses {
		if h.drink == "water" {
			waterDrinker = h.national
		}
		if h.pet == "zebra" {
			zebraOwner = h.national
		}
	}
	
	return Solution{
		DrinksWater: waterDrinker,
		OwnsZebra:   zebraOwner,
	}
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/cpp/exercises/practice/zebra-puzzle/zebra_puzzle.cpp
#include "zebra_puzzle.h"
#include <vector>
#include <algorithm>

using namespace std;

namespace zebra_puzzle {

// Helper function to generate permutations
template<typename T>
vector<vector<T>> permutations(const vector<T>& items) {
    vector<vector<T>> result;
    vector<T> copy(items);
    sort(copy.begin(), copy.end());
    do {
        result.push_back(copy);
    } while (next_permutation(copy.begin(), copy.end()));
    return result;
}

// Positional checks
bool RightOf(int a, int b) { return a == b + 1; }
bool NextTo(int a, int b) { return abs(a - b) == 1; }

Solution solve() {
    // Define all possible attributes
    const vector<string> nations = {"Englishman", "Spaniard", "Ukrainian", "Norwegian", "Japanese"};
    const vector<string> colors = {"Red", "Green", "Ivory", "Yellow", "Blue"};
    const vector<string> pets = {"Dog", "Snail", "Fox", "Horse", "Zebra"};
    const vector<string> drinks = {"Coffee", "Tea", "Milk", "OrangeJuice", "Water"};
    const vector<string> hobbies = {"Painting", "Dancing", "Reading", "Football", "Chess"};

    // Generate all possible permutations for each category
    for (const auto& nation : permutations(nations)) {
        // Clue 10: Norwegian lives in first house
        if (nation[0] != "Norwegian") continue;

        for (const auto& color : permutations(colors)) {
            // Clue 2: Englishman in Red house
            auto eng_house = find(nation.begin(), nation.end(), "Englishman") - nation.begin();
            if (color[eng_house] != "Red") continue;

            // Clue 6: Green house right of Ivory
            auto ivory = find(color.begin(), color.end(), "Ivory") - color.begin();
            auto green = find(color.begin(), color.end(), "Green") - color.begin();
            if (!RightOf(green, ivory)) continue;

            // Clue 15: Norwegian next to Blue
            auto blue = find(color.begin(), color.end(), "Blue") - color.begin();
            if (!NextTo(0, blue)) continue;

            for (const auto& drink : permutations(drinks)) {
                // Clue 9: Middle house drinks Milk
                if (drink[2] != "Milk") continue;

                // Clue 4: Green house drinks Coffee
                if (drink[green] != "Coffee") continue;

                // Clue 5: Ukrainian drinks Tea
                auto ukr = find(nation.begin(), nation.end(), "Ukrainian") - nation.begin();
                if (drink[ukr] != "Tea") continue;

                for (const auto& hobby : permutations(hobbies)) {
                    // Clue 8: Yellow house is painter
                    auto yellow = find(color.begin(), color.end(), "Yellow") - color.begin();
                    if (hobby[yellow] != "Painting") continue;

                    // Clue 13: Football player drinks OrangeJuice
                    auto football = find(hobby.begin(), hobby.end(), "Football") - hobby.begin();
                    if (drink[football] != "OrangeJuice") continue;

                    // Clue 14: Japanese plays Chess
                    auto japan = find(nation.begin(), nation.end(), "Japanese") - nation.begin();
                    if (hobby[japan] != "Chess") continue;

                    for (const auto& pet : permutations(pets)) {
                        // Clue 3: Spaniard owns Dog
                        auto spain = find(nation.begin(), nation.end(), "Spaniard") - nation.begin();
                        if (pet[spain] != "Dog") continue;

                        // Clue 7: Snail owner likes Dancing
                        auto snail = find(pet.begin(), pet.end(), "Snail") - pet.begin();
                        if (hobby[snail] != "Dancing") continue;

                        // Clue 11: Reading next to Fox
                        auto reading = find(hobby.begin(), hobby.end(), "Reading") - hobby.begin();
                        auto fox = find(pet.begin(), pet.end(), "Fox") - pet.begin();
                        if (!NextTo(reading, fox)) continue;

                        // Clue 12: Painter next to Horse
                        auto painter = find(hobby.begin(), hobby.end(), "Painting") - hobby.begin();
                        auto horse = find(pet.begin(), pet.end(), "Horse") - pet.begin();
                        if (!NextTo(painter, horse)) continue;

                        // Found valid solution
                        Solution sol;
                        auto water = find(drink.begin(), drink.end(), "Water") - drink.begin();
                        sol.drinksWater = nation[water];
                        auto zebra = find(pet.begin(), pet.end(), "Zebra") - pet.begin();
                        sol.ownsZebra = nation[zebra];
                        return sol;
                    }
                }
            }
        }
    }
    
    return {}; // Should never reach here with valid constraints
}

}  // namespace zebra_puzzle
